config/config.go

```diff
@@ -223,6 +239,104 @@ func (c *Config) Validate() error {
 	return nil
 }
 
+// If the numerator in the Warp config is 0, use the default value
+func setQuorumNumerator(cfgNumerator uint64) uint64 {
+	if cfgNumerator == 0 {
+		return params.WarpDefaultQuorumNumerator
+	}
+	return cfgNumerator
+}
+
+// Helper to retrieve the Warp Quorum from the chain config.
+// Differentiates between subnet-evm and coreth RPC internally
+func getWarpQuorum(
+	subnetID ids.ID,
+	client ethclient.Client,
+) (WarpQuorum, error) {
+	if subnetID == constants.PrimaryNetworkID {
+		return WarpQuorum{
+			QuorumNumerator:   params.WarpDefaultQuorumNumerator,
+			QuorumDenominator: params.WarpQuorumDenominator,
+		}, nil
+	} else {
+		// Fetch the subnet's chain config
+		chainConfig, err := client.ChainConfig(context.Background())
+		if err != nil {
+			return WarpQuorum{}, fmt.Errorf("failed to fetch chain config for subnet %s: %v", subnetID, err)
+		}
+
+		// First, check if the Warp precompile was enabled at genesis
+		warpConfig, ok := chainConfig.GenesisPrecompiles["warpConfig"].(*warp.Config)
+		if ok {
+			return WarpQuorum{
+				QuorumNumerator:   setQuorumNumerator(warpConfig.QuorumNumerator),
+				QuorumDenominator: params.WarpQuorumDenominator,
+			}, nil
+		}
+
+		// If we didn't find the Warp config in the genesis precompile list, check the upgrade precompiles
+		upgrades := chainConfig.ToWithUpgradesJSON().UpgradeConfig.PrecompileUpgrades
+		for _, precompile := range upgrades {
+			warpConfig, ok := precompile.Config.(*warp.Config)
+			if ok {
+				return WarpQuorum{
+					QuorumNumerator:   setQuorumNumerator(warpConfig.QuorumNumerator),
+					QuorumDenominator: params.WarpQuorumDenominator,
+				}, nil
+			}
+		}
+		return WarpQuorum{}, fmt.Errorf("failed to find warp config for subnet %s", subnetID)
```

Comment:
Is it possible for a subnet to change the warp quorum in an upgrade after it has already been set? If so, we would need to find the most recent value.

Reply:
Good call. We should be able to account for this by checking the upgrade config first, then the genesis config if necessary.

Reply:
Yeah, we should also account for multiple upgrades having different values, if that's possible. This could also get a little hairy if we're processing historical blocks over an upgrade where this value might change, but that is probably a pretty rare edgecase.

Reply:
Ah okay great, that helps a lot. Could you leave a comment in here somewhere explaining that?

Reply:
Sure thing, added